ocaml/xapi/xapi_tracked_user_agents.ml

```diff
@@ -0,0 +1,84 @@
+(*
+ * Copyright (C) Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+let max_user_agent_strlen = 64
+
+let max_num = 128
+
+let t = Hashtbl.create max_num
+
+let q = Queue.create ()
+
+let m = Mutex.create ()
+
+let with_lock = Xapi_stdext_threads.Threadext.Mutex.execute
+
+let update name version =
+  with_lock m @@ fun () ->
+  match Hashtbl.find_opt t name with
+  | Some v ->
+      if v <> version then Hashtbl.replace t name version
+  | None ->
+      ( if Queue.length q >= max_num then
+          let to_remove = Queue.pop q in
+          Hashtbl.remove t to_remove
+      ) ;
+      Queue.push name q ; Hashtbl.add t name version
+
+let get () = Hashtbl.to_seq t |> List.of_seq
```

Comment:
The structures are not thread-safe either. So the `get` needs to be protected as well.